modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/freelist/CacheFreeList.java

```diff
@@ -76,4 +86,188 @@ public CacheFreeList(
+            }
+
+            // The data is fragmented, so need to update the required fragment.
+            long nextLink = 0L;
+            boolean pageUpdated = false;
+            int scannedBytes = 0;
+            int updatedBytes = 0;
+
+            // We are under write lock here, so it is safe to read/write the first page.
+            boolean firstPage = true;
+
+            do {
+                final int cutItemId = firstPage ? itemId : PageIdUtils.itemId(nextLink);
+                final long curPageId = firstPage ? pageId : PageIdUtils.pageId(nextLink);
+
+                final long curPage = firstPage ? page : pageMem.acquirePage(grpId, curPageId, statHolder);
+
+                try {
+                    // It is safe to acquire a write lock on the next pages because we are under the write lock on the first page (head of
+                    // the entry). This means that reading of entry is not possible, since it requires holding a read lock on the first page
+                    // until all pages related to the entry are read. Moreover, following the current implementation of inserting a new
+                    // entry, the tail of the entry is always placed on empty pages so these pages cannot be used for other entries.
+                    long curPageAddr = firstPage ? pageAddr : pageMem.writeLock(grpId, curPageId, curPage);
+
+                    try {
+                        assert curPageAddr != 0L : "Failed to acquire write lock on the page [grpId=" + grpId + ", pageId=" + curPageId +
+                            ", page=" + curPage + ", firstPage=" + firstPage + ", nextLink=" + nextLink + ']';
+
+                        ByteBuffer buf = pageMem.pageBuffer(curPageAddr);
+
+                        DataPageIO curIo = DataPageIO.VERSIONS.forPage(curPageAddr);
+
+                        data = curIo.readPayload(curPageAddr, cutItemId, pageMem.realPageSize(grpId));
+
+                        buf.position(data.offset());
+                        buf.limit(data.offset() + data.payloadSize());
+
+                        // This variable contains the number of bytes that were updated on the current page.
+                        // Take into account that in the case when all bytes were updated, the method below returns Integer.MAX_VALUE.
+                        int updatedOnCurrPage = curIo.updateExpirationTimeFragmentData(
+                            row,
+                            buf,
+                            data.payloadSize(),
+                            updatedBytes,
+                            scannedBytes
+                        );
+
+                        pageUpdated = updatedOnCurrPage != 0;
+                        scannedBytes += data.payloadSize();
+                        updatedBytes += updatedOnCurrPage;
+
+                        if (pageUpdated && needWalDeltaRecord(pageId, page, walPlc)) {
+                            // TODO https://ggsystems.atlassian.net/browse/GG-38158
```

Comment:
We're not releasing new version without this fix, right? That would have been a very bad decision

Reply:
Well, it is a bug of course that should be fixed. However, this case is already broken because we do not log logical wal records for updated ttl, so anyway, it may be possible to lose such updates under some circumstances.
As a workaround, we can disable in-place updates of ttl for the case when persistence is enabled.